### PR TITLE
Refactor/dnas from scaffolder

### LINKED ISSUE #87: Refactor DNAs to follow patterns used in the scaffolder
- Separate concerns into different zomes
- Split large coodinator zome file into seperate files / zomes
- Attach links to base for faster querying of updates, deletes.

### LINKED ISSUE #98: Replace node test runner
Consider replacing node test runner we're currently using 'tape' with another that offers better output for failing tests.

Some options I like:
- ava: https://github.com/avajs/ava
- vitest: https://vitest.dev

### PR DESCRIPTION
**Pull request by mattyg**
Resolves #87 
Resolves #98 
Partially addresses #67

**Overview**
- re-generate all functionality using scaffolder, sticking to original generated patterns as much as possible
- split out 'likes' and 'follows' into their own zomes
- split out prefix indexing logic into its own crate
- split up test files by area of concern
- replace 'tap' test runner with vitest
- adds delete mew functionality to dna

**Outstanding issues**
- I think there are some limitations in validating links within a Path that I'm not sure how to resolve, not sure if you have any suggestions @jost-s  (see https://github.com/GeekGene/MewsFeed/pull/124/files#diff-3fa27f130d64c3bf52e39484f798ce224d90bb5744fd75e6edbb80d7fa6081f1R34)
- I want to validate that a link base is an AgentPubKey, ideally with `AgentPubKey::try_from(base_address)`, but we need holochain core to include AgentPubKey in AnyLinkableHash (see https://github.com/holochain/holochain/pull/2199)
- Also the way I implemented the "prefix index" was to make a seperate rust crate that can be imported into a zome. In this approach you can directly import the prefix index functions, pass in the index link type, and in your integrity zome specify that link type. An alternative pattern would be to make the prefix index a pair of zomes, with its own defined link type, and then have other coordinator zomes do a `call()` to interact with it. Anyone have any insights about which pattern might be preferable? @jost-s @harlantwood

### CODE REVIEW

**Review by jost-s — COMMENTED** [review]
When I create a mew with less than the minimum of characters, I'm allowed to submit it and the mewsfeed is reloaded, but nothing happens.
I suggest we block the "Publish Mew" button while the character count is below the minimum.

**Comment by jost-s** [comment]
Another suggestion, convert the "Mew Mentions" link to a button to stand out more
image

**Comment by jost-s** [comment]
> I think there are some limitations in validating links within a Path that I'm not sure how to resolve, not sure if you have any suggestions @jost-s

I'll have a look tomorrow

**Comment by bierlingm** [comment]
> When I create a mew with less than the minimum of characters, I'm allowed to submit it and the mewsfeed is reloaded, but nothing happens.
> 
> I suggest we block the "Publish Mew" button while the character count is below the minimum.

We have a minimum? What's the minimum? And what was the reasoning for a minimum? Spam prevention?

**Comment by mattyg** [comment]
> > When I create a mew with less than the minimum of characters, I'm allowed to submit it and the mewsfeed is reloaded, but nothing happens.
> > I suggest we block the "Publish Mew" button while the character count is below the minimum.
> 
> We have a minimum? What's the minimum? And what was the reasoning for a minimum? Spam prevention?

Yeah you can optionally specify a min and max character limits in DNA properties. It was implemented here: https://github.com/GeekGene/MewsFeed/pull/116

I don't really feel strongly about a rationale, I just figured while I'm adding a max limit might as well add a min limit too. It'll give different instances more dials to adjust. Maybe you're using it for organizations' weekly updates and you want to require at least 500 characters per update :shrug:

**Comment by mattyg** [comment]
> When I create a mew with less than the minimum of characters, I'm allowed to submit it and the mewsfeed is reloaded, but nothing happens. I suggest we block the "Publish Mew" button while the character count is below the minimum.

huh, its disabled on my browser. weird.

**Review by jost-s — APPROVED** [review]
I've tested the app by using it and browsed through the diff. Overall looks great to me!
I like the test upgrades and ran various tests - a good experience.

In terms of getting an agent pub key from an `AnyLinkableHash`, currently you can still do
```rust
let agent_pub_key = AgentPubKey::from(EntryHash::from(target_address));
```

That'll soon change in Holochain, but you'll still be able to do that conversion.

**Comment by jost-s** [comment]
> * Also the way I implemented the "prefix index" was to make a seperate rust crate that can be imported into a zome. In this approach you can directly import the prefix index functions, pass in the index link type, and in your integrity zome specify that link type. An alternative pattern would be to make the prefix index a pair of zomes, with its own defined link type, and then have other coordinator zomes do a `call()` to interact with it. Anyone have any insights about which pattern might be preferable?

It seems others could re-use the prefix index if it was its own DNA. That's preferable from my point of view, if it's possible to drop any link type into it as you say.

File: crates/prefix_index/src/validate.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    } else if EntryHash::from(base_address) == Path::from(index_name.into()).path_entry_hash()? {
+        // Second component
+        if tag_string.chars().count() != *PREFIX_INDEX_WIDTH {
+            return Ok(ValidateCallbackResult::Invalid(format!(
+                "PrefixIndex second component must be length {:?}",
+                *PREFIX_INDEX_WIDTH
+            )));
+        }
+    } else {
+        // Third component
```
Comment (jost-s) [inline crates/prefix_index/src/validate.rs:34]: I tried to reach this arm in my tests but it never got there - under which circumstances does that happen?

Comment (mattyg) [reply crates/prefix_index/src/validate.rs:34]: right-- if someone commits a PrefixIndex link of > 2 components it should just be invalid. thanks

**Comment by mattyg** [comment]
> > * Also the way I implemented the "prefix index" was to make a seperate rust crate that can be imported into a zome. In this approach you can directly import the prefix index functions, pass in the index link type, and in your integrity zome specify that link type. An alternative pattern would be to make the prefix index a pair of zomes, with its own defined link type, and then have other coordinator zomes do a `call()` to interact with it. Anyone have any insights about which pattern might be preferable?
> 
> It seems others could re-use the prefix index if it was its own DNA. That's preferable from my point of view, if it's possible to drop any link type into it as you say.

huh, you think it should be in it's own dna entirely? I was just thinking either publish it as a coordinator/integrity zome pair, *or* publish it as a rust crate that you import into your own zome pair. Either way the logic can be reused.

**Comment by harlantwood** [comment]
In case you want to grab it, this is the call_zome helper function I showed you:

https://github.com/GeekGene/MewsFeed/pull/104/files#diff-679fe327dd9f552952c6b6bfeb927d953326df5ebc24da4e47a96f6767a38238R679

**Comment by jost-s** [comment]
> huh, you think it should be in it's own dna entirely? I was just thinking either publish it as a coordinator/integrity zome pair, _or_ publish it as a rust crate that you import into your own zome pair. Either way the logic can be reused.

Yes, true. Between those two I don't have much of a preference. Slightly toward coordinator/integrity zome pair.

**Comment by harlantwood** [comment]
Congrats!  Huge effort, well done!  🎉👏

**Pull request merged**